sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,513 @@
+		Report:       report,
+		AttKeyId:     attKeyId,
+		QeReportInfo: nil,
+		BufSize:      quoteSize,
+		Timeout:      aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetQuoteExRes = &pb.AesmServiceResponse_GetQuoteExResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, err
+	}
+
+	if errCode := resp.GetQuoteExRes.GetErrorCode(); errCode != 0 {
+		return nil, fmt.Errorf("failed to call getQuoteExRequest (error code = %d)", errCode)
+	}
+
+	quote := resp.GetQuoteExRes.GetQuote()
+	if quote == nil {
+		return nil, fmt.Errorf("failed to call getQuoteExRequest")
+	}
+
+	if len(quote) != (int)(quoteSize) {
+		return nil, fmt.Errorf("failed to call getQuoteExRequest: len(quote) is not %d but %d\n",
+			quoteSize, len(quote))
+	}
+
+	qeReportInfo := resp.GetQuoteExRes.GetQeReportInfo()
+	if qeReportInfo != nil {
+		return nil, fmt.Errorf("failed to call getQuoteExRequest: invalid Qe Report Info:%v\n", qeReportInfo)
+	}
+
+	return quote, nil
+}
+
+func selectAttKeyIDRequest() ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.SelectAttKeyIDReq = &pb.AesmServiceRequest_SelectAttKeyIDRequest{
+		Timeout: aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.SelectAttKeyIDRes = &pb.AesmServiceResponse_SelectAttKeyIDResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, err
+	}
+
+	if errCode := resp.SelectAttKeyIDRes.GetErrorCode(); errCode != 0 {
+		return nil, fmt.Errorf("failed to call selectAttKeyIDRequest (error code = %d)", errCode)
+	}
+
+	akId := resp.SelectAttKeyIDRes.GetSelectedAttKeyId()
+	if akId == nil {
+		return nil, fmt.Errorf("failed to call selectAttKeyIDRequest")
+	}
+
+	return akId, nil
+}
+
+func getSupportedAttKeyIDNumRequest() (uint32, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return 0, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.GetSupportedAttKeyIDNumReq = &pb.AesmServiceRequest_GetSupportedAttKeyIDNumRequest{
+		Timeout: aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return 0, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetSupportedAttKeyIDNumRes = &pb.AesmServiceResponse_GetSupportedAttKeyIDNumResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return 0, err
+	}
+
+	if errCode := resp.GetSupportedAttKeyIDNumRes.GetErrorCode(); errCode != 0 {
+		return 0, fmt.Errorf("failed to call getSupportedAttKeyIDNumRequest (error code = %d)", errCode)
+	}
+
+	akIdNum := resp.GetSupportedAttKeyIDNumRes.GetAttKeyIdNum()
+	if akIdNum == 0 {
+		return 0, fmt.Errorf("failed to call getSupportedAttKeyIDNumRequest: invalid Attestation Key Identity Number = %d", akIdNum)
+	}
```

Comment:
Yes, returning zero akId implies something goes wrong, but checking this situation is not the business of a thin wrapper function. Move the sanity check to the caller.